Source/WebCore/dom/Element.cpp

```diff
@@ -2498,6 +2499,52 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
             }
         }
     }
+
+    if (langAttrMatchesDocument()) {
+        oldDocument.removeElementWithLangAttrMatchingDocument(*this);
+        setLangAttrMatchesDocument(false);
+    }
+
+    setEffectiveLang();
+}
+
+void Element::setEffectiveLangFromParent()
+{
+    ASSERT(langFromAttribute().isNull());
+
+    if (auto* parent = parentOrShadowHostElement(); parent && parent != document().documentElement()) {
+        setEffectiveLangMatchesDocument(parent->effectiveLangMatchesDocument());
+        if (UNLIKELY(parent->hasRareData()) && !parent->elementRareData()->effectiveLang().isNull())
+            ensureElementRareData().setEffectiveLang(parent->elementRareData()->effectiveLang());
+        else if (hasRareData())
+            elementRareData()->setEffectiveLang(nullAtom());
+    }
+}
+
+void Element::setEffectiveLang()
+{
+    auto& lang = langFromAttribute();
```

Comment:
Hm... this would mean every adoption of an element results in O(k) traversal where k is the number of attributes.

Can we instead remember if this element has an explicit lang attribute set or not using a EventTarget bit flag? Then the combination of that flag plus the flag indicating that the effective lang matches that of the document element should give us logically equivalent sets of flags.

Reply:
Unfortunately we'd still need an extra bit, since we'd need to track whether the element has a `lang` and an `xml:lang`. Unless upon removal of one we search for whether the other one is present.